src/Conversion/ONNXToTOSA/Tensor/Resize.cpp

```diff
@@ -0,0 +1,332 @@
+
+  // If round_half_up we need to adjust the offset
+  if (isNearest && isNearestModeFloor) {
+    offset -= numerator / 2;
+  }
+
+  // We can compute this directly based on previous values.
+  border = denominator * (output - 1) - numerator * (input - 1) + offset;
+  return ScaleHelper(numerator, denominator, offset, border);
+};
+
+void valuesFromAxis(ArrayAttr *axis, llvm::SmallVectorImpl<int64_t> &axisVec) {
+  auto axisRange = axis->getAsRange<IntegerAttr>();
+  llvm::transform(axisRange, std::back_inserter(axisVec), [](IntegerAttr attr) {
+    return tosa::convertNegativeAxis(attr.getInt(), 4);
+  });
+}
+
+LogicalResult getScaleValue(ConversionPatternRewriter &rewriter, Operation *op,
+    llvm::SmallVectorImpl<int64_t> &axisVec,
+    llvm::SmallVectorImpl<float> &scaleVec, Value scaleValue) {
+  mlir::ElementsAttr elementsAttr =
+      getElementAttributeFromONNXValue(scaleValue);
+  if (!elementsAttr)
+    return rewriter.notifyMatchFailure(
+        op, "Scale cannot come from a block argument.");
+
+  // The axis attribute might permute and/or reduce the number of elements
+  // in scaleValue. This reorders the scales to account for that.
+  for (auto [index, value] : llvm::enumerate(axisVec))
+    scaleVec[value] =
+        elementsAttr.getValues<FloatAttr>()[index].getValueAsDouble();
+
+  // Even if the shapes are identical, a scale value other than 1 is
+  // possible. TOSA does not allow that for non-spatial dimensions.
+  for (int64_t nonSpatialDimensions : {0, 1}) {
+    if (scaleVec[nonSpatialDimensions] != 1)
+      return rewriter.notifyMatchFailure(
+          op, "Axis Attr if present must contain both output dimensions.");
+  }
+  return success();
+}
+
+class ONNXResizeOpLoweringToTOSA : public ConversionPattern {
+public:
+  ONNXResizeOpLoweringToTOSA(MLIRContext *ctx)
+      : ConversionPattern(ONNXResizeOp::getOperationName(), 1, ctx) {}
+  using OpAdaptor = typename ONNXResizeOp::Adaptor;
+
+  struct FractionNumber {
+    FractionNumber(float number) {
+      double integral = std::floor(number);
+      double frac = number - integral;
+
+      const long precision = 1000000000; // This is the accuracy.
+
+      long gcd_ = std::gcd((int)round(frac * precision), precision);
+
+      long denominator = precision / gcd_;
+      long numerator = round(frac * precision) / gcd_;
+
+      this->numerator = numerator + denominator * integral;
+      this->denominator = denominator;
+    }
+    int64_t numerator;
+    int64_t denominator;
+  };
+
+  /// ## coordinateTransformationMode ##
+  /// TOSA uses the formula ix = (ox * scale_x_d + offset_x) / scale_x_n
+  /// to find the input coordinates. In order to lower ONNX one needs to
+  /// express the modes in this context. Border is only used to check if certain
+  /// conditions in the dimensions are met, but has no actual use in calculating
+  /// something. It is probably an error in the TOSA specification. In order to
```

Comment:
Border is used to ensure that the shape calculation is exactly defined with an integer ratio.